File: go/src/dataprocessor/resultsaver.go
package main // Important: Belongs to the main package now

import (
    "os"
)

// ResultSaver struct to hold file writer.
type ResultSaver struct {
    fileWriter *os.File
    filename   string
}

// NewResultSaver creates a new ResultSaver.
func NewResultSaver(filename string) (*ResultSaver, error) {
    file, err := os.Create(filename)
    if err != nil {
        return nil, err
    }
    return &ResultSaver{fileWriter: file, filename: filename}, nil
}

// SaveResult saves a result to the file.
func (rs *ResultSaver) SaveResult(result string) error {
    _, err := rs.fileWriter.WriteString(result + "\n")
    if err != nil {
        return err
    }
    return nil
}

// Close closes the file writer.
func (rs *ResultSaver) Close() error {
    if rs.fileWriter != nil {
        return rs.fileWriter.Close()
    }
    return nil
}

File: java/src/main/java/com/example/ResultSaver.java
package com.example;

import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReentrantLock;

public class ResultSaver {
    private List<String> results = new ArrayList<>();
    private Lock lock = new ReentrantLock();
    private FileWriter fileWriter;
    private String filename;

    public ResultSaver(String filename) throws IOException {
        // Construct the file path using the provided relative path "java/results"
        String basePath = new File("").getAbsolutePath();
        System.out.println("hello world" + basePath);
        this.filename = basePath + File.separator + results + File.separator + filename;
        System.out.println(this.filename);

        // Ensure the directory exists
        File file = new File(this.filename);
        file.getParentFile().mkdirs();
        this.fileWriter = new FileWriter(file);
    }

    public void saveResult(String result) {
        lock.lock();
        try {
            results.add(result);
            fileWriter.write(result + "\n");
        } catch (IOException e) {
            System.err.println("Error writing to file: " + e.getMessage());
        } finally {
            lock.unlock();
        }
    }

    public void close() throws IOException {
        fileWriter.close();
    }
}
